### PR TITLE
Clustering coefficient

### PR DESCRIPTION
**Pull request by AlexandruValeanu**
Following the discussions from #607, I've added an algorithm for computing local/glocal clustering coefficients.

----

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/ClusteringCoefficient.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    private void computeGlobalClusteringCoefficient(){
+        computed = true;
+        double numberTriplets = 0;
+
+        for (V v: graph.vertexSet()){
+            if (graph.getType().isUndirected()){
+                numberTriplets += 1.0 * graph.degreeOf(v) * (graph.degreeOf(v) - 1) / 2;
+            }
+            else{
+                numberTriplets += 1.0 * graph.inDegreeOf(v) * graph.outDegreeOf(v);
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/scoring/ClusteringCoefficient.java]: Is this valid? For two vertices, 1 and 2, there may be two arcs (1,2) and (2,1). Yet this is not a triplet.

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/ClusteringCoefficient.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+    private void computeScoreMap(){
+        if (scores != null)
+            return;
+
+        scores = new HashMap<>(graph.vertexSet().size());
+
+        for (V v: graph.vertexSet()){
+            Set<V> neighbourhood = new HashSet<>();
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/scoring/ClusteringCoefficient.java]: You might be able to reuse the class NeighborhoodCache for this.

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/ClusteringCoefficient.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+     * @param v the vertex
+     * @return the local clustering coefficient
+     */
+    @Override
+    public Double getVertexScore(V v) {
+        if (!graph.containsVertex(v)) {
+            throw new IllegalArgumentException("Cannot return score of unknown vertex");
+        }
+
+        computeScoreMap();
```
Comment (assimiz) [inline jgrapht-core/src/main/java/org/jgrapht/alg/scoring/ClusteringCoefficient.java]: Why do we compute the complete score map when a single vertex score is requested? It is not always the case that we would want the average/global coefficient.

Comment (AlexandruValeanu) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/ClusteringCoefficient.java]: You are right. There's no need to computed the whole map. I've changed it to do the computation per vertex when needed.

**Comment by jkinable** [comment]
Looks fine to me.
@d-michail any additional remarks or shall I go ahead and merge this one?

**Comment by d-michail** [comment]
Looks fine

**Comment by jkinable** [comment]
thanks!

**Pull request merged**